scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,289 @@
+"""Main class and helper functions."""
+import inspect
+import logging
+import os
+import warnings
+from typing import List, Optional, Union
+
+import anndata
+import numpy as np
+import ray
+import torch
+from ray import tune
+from ray.tune import CLIReporter, ExperimentAnalysis
+from ray.tune.schedulers import ASHAScheduler
+from ray.tune.schedulers.trial_scheduler import TrialScheduler
+
+from scvi.model.base import BaseModelClass
+
+from ._utils import apply_model_config, fetch_config, train_model
+
+logger = logging.getLogger(__name__)
+
+
+class Autotune:
+    """
+
+    Hyperparameter tuning using Ray Tune.
+
+    Parameters
+    ----------
+    adata
+        AnnData object we will tune the model on.
+    model
+        Model from scvi.model we will tune.
+    num_epochs
+        Number of epochs to tune the model over
+    training_metrics
+        Metrics to track during training.
+    metric_functions
+        For metrics calculated after training a model, like silhouette distance.
+    model_hyperparams
+        Config for the model hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    trainer_hyperparams
+        Config for the trainer hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    plan_hyperparams
+        Config for the training_plan hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    setup_args
+        Non tunable anndata setup parameters
+    test_effect_covariates
+        Whether to test the effect of including covariates in the model. Default is `True`
+    continuous_covariates
+        continuous covariates from `adata` to tune the model on
+    categorical_covariates
+        Categorical covariates from `adata` to tune the model on
+    test_effect_hvg
+        Whether to test the effect of filtering adata with highly variable genes on model performance or not. Default is `False`.
+    top_hvg
+        Top hvgs to test. If `test_effect_hvg` is set, defaults to top 2000, 2500, 3000 and 3500 highly variable genes
+    batch_key_hvg
+        Column in adata.obs specifying the different batches in the data used for highly_variable_gene selection
```

Comment:
I think this is a bit overloaded. I think this API would be easier to use if Autotune did not alter the data/ setup anndata hyperparams at all, because one can always run multiple autotune sessions for different data configs. These kwargs are a bit too specific to scvi as it currently is.

An alternative is to provide a function that alters the data as input and a config or something, but I think we can leave that for future work

Reply:
Therefore I think we should simplify this functionality and only have setup_kwargs. Wyt? @PierreBoyeau @jjhong922

Reply:
Indeed. As it is, it might be too much functionality packed in the class initialization. However, I think the ability to test different anndata configurations is quite handy, so your suggestion to factor it out as a separate function can be the sweet spot, maybe something like `.test_configurations()`? 

Let me know what you all think about this, so I can factor it out and put it as a separate block of code for future use.